src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -4132,7 +4132,8 @@ trait Types
       while (tparams1 != tparams0) {
         tparams0 = tparams1
         tparams1 = tparams filter { p =>
```

Comment:
🤔 should this be `tparams1 = tparams0 filter ...`  (I know this was already there)

smells like an opportunity for redoing this with more fitting data structures?

Reply:
When I saw this code I found it quite intriguing, and I had doubts as well. But no, it is `tparams1` that it should use.

This is a fixed-point: the `tparams0` is only used to check, by comparison, if the "selected" set `tparams1` has changed. The idea of `tparams1` is to keep a _frontier_, a subset of the input `tparams` that are already known to be directly or indirectly referenced from `tpe`.

There may be an opportunity to redo this with more fitting data structures, yes, and this code seems to may redundant checks. However, I do not know how much relevant this code weighs on overall performance, to use anything more than two lines.

Not that I did not try, though. I will push a different branch with a more elaborate change.

Reply:
but how about the RHS? Why is it filtering `tparams`, and not `tparams0` each time, which doesn't change?

Reply:
The key point is that this is a transitive closure: later iterations of the loop will allow _more_ elements from `tparams` into `tparams1`. If the order of the `tparams` in the final list is not a problem (and if it is an existential it ideally should not be), then we could improve this.

Reply:
This is the branch where I have made more elaborate changes. The only problem is that `partest` fails.
 https://github.com/scala/scala/compare/2.13.x...diesalbla:existential_abstraction_loops?expand=1

Reply:
> but how about the RHS? Why is it filtering tparams, and not tparams0 each time. tparams doesn't change, while tparams0 does.

I think t would have been more clear if `tparams1` and `tparams0` had been called `existParams` and `existParamsPrev`, respectively.

Reply:
That makes sense. Here's what I came up with (running test suite now):

```
      val tpe1     = extrapolation extrapolate tpe

      val (tparamsInitial, tparamsRemaining) = tparams.partition(tpe1.contains)
      val tparamClosure = mutable.ListBuffer.empty[Symbol] ++ tparamsInitial
      val tparamsTodo   = mutable.ListBuffer.empty[Symbol] ++ tparamsRemaining

      var oldLen = 0
      while (tparamsTodo.nonEmpty && oldLen != tparamClosure.length) {
        oldLen = tparamClosure.length
        tparamsTodo.filterInPlace { paramTodo =>
          // If our closure is not yet complete (there's a tparam in tparamsClosed whose info refers to a tparam in paramTodo),
          // add `paramTodo` to `tparamClosure`, and drop it from `tparamsTodo`
          !tparamClosure.exists(_.info contains paramTodo) || { tparamClosure += paramTodo; false }
        }
      }

      newExistentialType(tparamClosure.toList, tpe1)
```